File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.py
from collections import deque
def dfs(graph, v, visited):
    # 현재 노드를 방문 처리
    visited[v] = True
    print(v, end=' ')

    for i in graph[v]:
        if not visited[i]:
            dfs(graph, i, visited)

def bfs(graph, start, visited):
    queue = deque([start])
    visited[start] = True

    while queue:
        v = queue.popleft()
        print(v, end = ' ')
        for i in graph[v]:
            if not visited[i]:
                queue.append(i)
                visited[i] = True

N, M, V = map(int, input().split())

visited = [False] * (N+1)
graph2 = [[] for i in range(N+1)]

for _ in range(M):
    a,b=map(int,input().split())
    graph2[a].append(b)
    graph2[b].append(a)

for i in range(len(graph2)):
    graph2[i].sort()

dfs(graph2, V, visited)
print()
visited = [False] * (N+1)
bfs(graph2,V,visited)

File: 백준/Silver/1260. DFS와 BFS/DFS와 BFS.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    static int N, M, V;
    static boolean[] visited;
    static int[][] graph;
    static StringBuilder sb;

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        N = Integer.parseInt(st.nextToken());
        M = Integer.parseInt(st.nextToken());
        V = Integer.parseInt(st.nextToken());

        graph = new int[N + 1][N + 1];

        for (int i = 0; i < M; i++) {
            st = new StringTokenizer(br.readLine());
            int s = Integer.parseInt(st.nextToken());
            int e = Integer.parseInt(st.nextToken());
            graph[s][e] = graph[e][s] = 1;
        }

        sb = new StringBuilder();
        visited = new boolean[N + 1];
        dfs(V);
        System.out.println(sb);
        sb = new StringBuilder();
        visited = new boolean[N + 1];
        bfs(V);
        System.out.println(sb);
    }


    static void dfs(int start) {
        visited[start] = true;
        sb.append(start + " ");
        for (int i = 1; i < N + 1; i++) {
            if (graph[start][i] == 1 && visited[i] != true) {
                dfs(i);
            }
        }
    }

    static void bfs(int start) {
        Queue<Integer> q = new LinkedList<>();
        q.add(start);
        visited[start] = true;

        while (!q.isEmpty()) {
            int node = q.poll();
            sb.append(node + " ");

            for (int i = 1; i < N + 1; i++) {
                if (graph[node][i] == 1 && visited[i] != true) {
                    visited[i] = true;
                    q.add(i);
                }
            }
        }
    }
}
